File: scripts/build_2025_inserts.js
#!/usr/bin/env node
// Parses 24 downloaded Sheets JSON files (KZ + KG, 12 months of 2025)
// from C:/Users/Мирзахит/Downloads/p2025/<KZ|KG>_<Янв..Дек>.json
// Emits a single SQL file with multi-row INSERTs into public.payments
// with ON CONFLICT-safe duplicate guard.
//
// Output: C:/Users/Мирзахит/Downloads/p2025/payments_2025_inserts.sql
//
// Run: node scripts/build_2025_inserts.js

const fs = require('fs');
const path = require('path');

const SRC_DIR = 'C:/Users/Мирзахит/Downloads/p2025';
const OUT_FILE = 'C:/Users/Мирзахит/Downloads/p2025/payments_2025_inserts.sql';
const STATS_FILE = 'C:/Users/Мирзахит/Downloads/p2025/payments_2025_stats.json';

const MONTHS = ['Янв','Фев','Мар','Апр','Май','Июн','Июл','Авг','Сен','Окт','Ноя','Дек'];

const COUNTRIES = {
  KZ: {
    sheet_id: '11ErpSR9fJ_T0ggWBrHjRB35cMs4yl84HedSO1Tf4Z08',
    currency: 'KZT',
    dateCorrection: 0
  },
  KG: {
    sheet_id: '1e34EE4DKuj2tzatlX1qBFEAEX7aDtky02R2HA2KE7_M',
    currency: 'KGS',
    dateCorrection: 1
  }
};

// ---- helpers (cloned 1:1 from api/payments.js) ----

function parseDate(v, dateCorrection) {
  if (v == null) return null;
  if (typeof v === 'number') {
    if (v === 0) return null;
    const d = new Date(Math.round((v - 25569) * 86400 * 1000));
    if (isNaN(d)) return null;
    return d.getFullYear() + '-' + String(d.getMonth() + 1).padStart(2, '0') + '-' + String(d.getDate()).padStart(2, '0');
  }
  let s = String(v).trim();
  if (!s || s === '0' || s === '-' || s === '—') return null;
  const m1 = s.match(/Date\((\d+),(\d+),(\d+)\)/);
  if (m1) return m1[1] + '-' + String(parseInt(m1[2]) + 1).padStart(2, '0') + '-' + String(m1[3]).padStart(2, '0');
  const m2 = s.match(/^(\d{1,2})\.(\d{1,2})\.(\d{4})/);
  if (m2) {
    const dObj = new Date(parseInt(m2[3]), parseInt(m2[2]) - 1, parseInt(m2[1]) + (dateCorrection || 0));
    if (isNaN(dObj.getTime())) return null;
    return dObj.getFullYear() + '-' + String(dObj.getMonth() + 1).padStart(2, '0') + '-' + String(dObj.getDate()).padStart(2, '0');
  }
  if (s.match(/^\d{4}-\d{2}-\d{2}/)) return s.substring(0, 10);
  try {
    const d = new Date(s);
    if (!isNaN(d) && d.getFullYear() > 2000) {
      return d.getFullYear() + '-' + String(d.getMonth() + 1).padStart(2, '0') + '-' + String(d.getDate()).padStart(2, '0');
    }
  } catch (e) {}
  return null;
}

function mapCategory(cat) {
  const c = String(cat || '').toLowerCase().replace(/\./g, ' ').replace(/\s+/g, ' ').trim();
  if (c.includes('интеграц')) return 'integration';
  if (c.includes('внедрен') || c.includes('доработ')) return 'implementation';
  if (c.includes('абон') || c.includes('баланс')) return 'subscription';
  if (c.includes('лицен') || c.includes('новый клиент') || c.includes('нов клиент')) return 'license';
  return 'other';
}

function parseNumber(v) {
  if (v == null) return null;
  if (typeof v === 'number') return v;
  const s = String(v).replace(/\s/g, '').replace(/[^0-9.,-]/g, '').replace(',', '.');
  const n = parseFloat(s);
  return isNaN(n) ? null : n;
}

function parseRow(row, hdrRow, cfg, monthName, monthIdx, sheetRowAbs) {
  if (!row || row[1] == null || String(row[1]).trim() === '') return null;
  let colDate = 0, colClient = 1, colCat = 2, colMgr = 5, colAmt = 12, colBank = 10, colSeated = 11, colTech = 17, colActivation = -1;
  if (hdrRow) {
    hdrRow.forEach((h, i) => {
      const s = String(h || '').toLowerCase();
      if (s === 'дата' && i < 3) colDate = i;
      else if (s === 'компания') colClient = i;
      else if (s.includes('стать')) colCat = i;
      else if (s === 'менеджер') colMgr = i;
      else if (s.includes('сумма') && s.includes('факт')) colAmt = i;
      else if (s.includes('сумма') && !s.includes('факт') && !s.includes('оста')) { if (colAmt === 9) colAmt = i; }
      else if (s === 'банк') colBank = i;
      else if (s.includes('посаж')) colSeated = i;
      else if (s.includes('тех') && !s.includes('актив')) colTech = i;
    });
    const h14 = String(hdrRow[14] || '').toLowerCase();
    if (h14.includes('актив')) colActivation = 14;
    else hdrRow.forEach((h, i) => {
      const s = String(h || '').toLowerCase();
      if (colActivation < 0 && s.includes('дата') && s.includes('актив') && !s.includes('цена')) colActivation = i;
    });
    if (colActivation < 0) colActivation = 14;
  }
  let amtRaw = row[colAmt];
  if (amtRaw == null || String(amtRaw).trim() === '') amtRaw = row[9];
  const amount = parseNumber(amtRaw);
  if (!amount || amount <= 0) return null;
  let paidAt = parseDate(row[colDate], cfg.dateCorrection);
  if (!paidAt) paidAt = `2025-${String(monthIdx + 1).padStart(2, '0')}-01`;
  const company = String(row[colClient] || '').trim();
  if (!company) return null;
  const catRaw = String(row[colCat] || '').trim();
  let actDate = parseDate(row[colActivation], cfg.dateCorrection);
  if (!actDate && colActivation === 14) actDate = parseDate(row[15], cfg.dateCorrection);
  return {
    paid_at: paidAt,
    company_name: company,
    category_raw: catRaw,
    category: mapCategory(catRaw),
    amount,
    amount_planned: parseNumber(row[9]),
    manager_name: String(row[colMgr] || '').trim() || null,
    bank: String(row[colBank] || '').trim() || null,
    seated: (function () { const v = String(row[colSeated] || '').trim(); return v === 'Да' || v === 'да' || v === '+'; })(),
    tech_support: String(row[colTech] || '').trim() || null,
    qty: parseNumber(row[4]),
    price: parseNumber(row[7]),
    period_months: (function () { const n = parseNumber(row[8]); return n != null ? Math.max(0, Math.min(60, Math.round(n))) : null; })(),
    activation_date: actDate,
    period_start_raw: row[19] ? String(row[19]).trim() : null,
    sheet_tab: monthName,
    sheet_row: sheetRowAbs
  };
}

function sqlStr(v) {
  if (v === null || v === undefined) return 'NULL';
  if (typeof v === 'number') return String(v);
  if (typeof v === 'boolean') return v ? 'true' : 'false';
  return "'" + String(v).replace(/'/g, "''") + "'";
}

// ---- main ----

const allRows = [];
const stats = { by_country: {}, by_month: {} };

for (const country of Object.keys(COUNTRIES)) {
  const cfg = COUNTRIES[country];
  stats.by_country[country] = { rows: 0, sum: 0 };
  for (let mi = 0; mi < MONTHS.length; mi++) {
    const monthName = MONTHS[mi];
    const file = path.join(SRC_DIR, `${country.toLowerCase()}_${monthName}.json`);
    if (!fs.existsSync(file)) { console.warn('MISS', file); continue; }
    let raw;
    try { raw = fs.readFileSync(file, 'utf8'); }
    catch (e) { console.warn('READ ERR', file, e.message); continue; }
    let data;
    try { data = JSON.parse(raw); }
    catch (e) { console.warn('JSON ERR', file, e.message); continue; }
    if (!data || !Array.isArray(data.rows) || data.rows.length < 2) { continue; }
    // Find header row (the one where row[1] === 'Компания')
    let headerIdx = -1;
    for (let ri = 0; ri < data.rows.length; ri++) {
      const r = data.rows[ri];
      if (r && String(r[1] || '').trim() === 'Компания') { headerIdx = ri; break; }
    }
    const hdrRow = headerIdx >= 0 ? data.rows[headerIdx] : null;
    const startIdx = headerIdx >= 0 ? headerIdx + 1 : 4;
    let monthRows = 0, monthSum = 0;
    data.rows.slice(startIdx).forEach((row, idx) => {
      const sheetRowAbs = startIdx + idx + 1;
      const parsed = parseRow(row, hdrRow, cfg, monthName, mi, sheetRowAbs);
      if (!parsed) return;
      parsed.country = country;
      parsed.currency = cfg.currency;
      parsed.source = 'sheets_import';
      parsed.sheet_id = cfg.sheet_id;
      // Force 2025 if for some reason date got parsed to 2026
      if (!/^2025-/.test(parsed.paid_at)) {
        const d = new Date(parsed.paid_at);
        if (!isNaN(d) && d.getFullYear() !== 2025) {
          parsed.paid_at = '2025-' + parsed.paid_at.substring(5);
        }
      }
      allRows.push(parsed);
      monthRows++; monthSum += parsed.amount;
    });
    stats.by_country[country].rows += monthRows;
    stats.by_country[country].sum += monthSum;
    stats.by_month[`${country}_${monthName}`] = { rows: monthRows, sum: monthSum };
  }
}

// ---- emit SQL ----

const BATCH = 100;
const cols = [
  'country','paid_at','company_name','category','category_raw','amount','amount_planned',
  'currency','qty','price','period_months','bank','manager_name','tech_support','seated',
  'activation_date','period_start_raw','source','sheet_id','sheet_tab','sheet_row'
];

const lines = [];
lines.push('-- Auto-generated 2025 payments import for KZ+KG');
lines.push(`-- Total rows: ${allRows.length}`);
lines.push(`-- Per-country: ${JSON.stringify(stats.by_country)}`);
lines.push('');
lines.push('BEGIN;');
lines.push('');

for (let i = 0; i < allRows.length; i += BATCH) {
  const slice = allRows.slice(i, i + BATCH);
  lines.push(`INSERT INTO payments (${cols.join(', ')}) VALUES`);
  const valueRows = slice.map(p => {
    const vals = [
      sqlStr(p.country),
      sqlStr(p.paid_at),
      sqlStr(p.company_name),
      sqlStr(p.category),
      sqlStr(p.category_raw),
      sqlStr(p.amount),
      sqlStr(p.amount_planned),
      sqlStr(p.currency),
      sqlStr(p.qty),
      sqlStr(p.price),
      sqlStr(p.period_months),
      sqlStr(p.bank),
      sqlStr(p.manager_name),
      sqlStr(p.tech_support),
      sqlStr(p.seated),
      sqlStr(p.activation_date),
      sqlStr(p.period_start_raw),
      sqlStr(p.source),
      sqlStr(p.sheet_id),
      sqlStr(p.sheet_tab),
      sqlStr(p.sheet_row)
    ];
    return `  (${vals.join(', ')})`;
  });
  lines.push(valueRows.join(',\n'));
  lines.push('  ON CONFLICT DO NOTHING;');
  lines.push('');
}

lines.push('COMMIT;');
lines.push('');

fs.writeFileSync(OUT_FILE, lines.join('\n'), 'utf8');
fs.writeFileSync(STATS_FILE, JSON.stringify({ total_rows: allRows.length, ...stats }, null, 2), 'utf8');

console.log('OK rows=' + allRows.length, ' file=' + OUT_FILE);
console.log('stats=', JSON.stringify(stats.by_country));

File: scripts/build_2025_inserts.py
#!/usr/bin/env python3
"""Parses 24 Sheets JSON files (KZ + KG, 12 months of 2025)
and emits a single SQL file with INSERTs into public.payments.

Output:
  C:/Users/Мирзахит/Downloads/p2025/payments_2025_inserts.sql
  C:/Users/Мирзахит/Downloads/p2025/payments_2025_stats.json
"""
import json, os, re, sys
from datetime import date

SRC_DIR = r'C:/Users/Мирзахит/Downloads/p2025'
OUT_SQL = os.path.join(SRC_DIR, 'payments_2025_inserts.sql')
OUT_STATS = os.path.join(SRC_DIR, 'payments_2025_stats.json')

MONTHS = ['Янв','Фев','Мар','Апр','Май','Июн','Июл','Авг','Сен','Окт','Ноя','Дек']

COUNTRIES = {
    'KZ': {'sheet_id': '11ErpSR9fJ_T0ggWBrHjRB35cMs4yl84HedSO1Tf4Z08', 'currency': 'KZT', 'date_corr': 0},
    'KG': {'sheet_id': '1e34EE4DKuj2tzatlX1qBFEAEX7aDtky02R2HA2KE7_M', 'currency': 'KGS', 'date_corr': 1},
}


def parse_date(v, date_corr):
    if v is None: return None
    if isinstance(v, (int, float)):
        if v == 0: return None
        # Excel serial -> JS-like: (v-25569)*86400 sec since epoch
        import datetime
        try:
            d = datetime.datetime(1970,1,1) + datetime.timedelta(seconds=int(round((v-25569)*86400)))
            return d.strftime('%Y-%m-%d')
        except Exception:
            return None
    s = str(v).strip()
    if not s or s in ('0','-','—'): return None
    m1 = re.match(r'Date\((\d+),(\d+),(\d+)\)', s)
    if m1:
        return f"{m1.group(1)}-{int(m1.group(2))+1:02d}-{int(m1.group(3)):02d}"
    m2 = re.match(r'^(\d{1,2})\.(\d{1,2})\.(\d{4})', s)
    if m2:
        import datetime
        try:
            d = datetime.date(int(m2.group(3)), int(m2.group(2)), 1) + datetime.timedelta(days=int(m2.group(1))-1 + (date_corr or 0))
            return d.strftime('%Y-%m-%d')
        except Exception:
            return None
    if re.match(r'^\d{4}-\d{2}-\d{2}', s):
        return s[:10]
    return None


def map_category(cat):
    c = re.sub(r'\s+', ' ', re.sub(r'\.', ' ', str(cat or '').lower())).strip()
    if 'интеграц' in c: return 'integration'
    if 'внедрен' in c or 'доработ' in c: return 'implementation'
    if 'абон' in c or 'баланс' in c: return 'subscription'
    if 'лицен' in c or 'новый клиент' in c or 'нов клиент' in c: return 'license'
    return 'other'


def parse_num(v):
    if v is None: return None
    if isinstance(v, (int, float)): return v
    s = re.sub(r'[^0-9.,-]', '', re.sub(r'\s', '', str(v))).replace(',', '.')
    try:
        return float(s)
    except Exception:
        return None


def parse_row(row, hdr_row, cfg, month_name, month_idx, sheet_row_abs):
    if not row or len(row) < 2 or row[1] is None or str(row[1]).strip() == '':
        return None
    col_date, col_client, col_cat, col_mgr = 0, 1, 2, 5
    col_amt, col_bank, col_seated, col_tech = 12, 10, 11, 17
    col_activation = -1
    if hdr_row:
        for i, h in enumerate(hdr_row):
            s = str(h or '').lower()
            if s == 'дата' and i < 3: col_date = i
            elif s == 'компания': col_client = i
            elif 'стать' in s: col_cat = i
            elif s == 'менеджер': col_mgr = i
            elif 'сумма' in s and 'факт' in s: col_amt = i
            elif 'сумма' in s and 'факт' not in s and 'оста' not in s:
                if col_amt == 9: col_amt = i
            elif s == 'банк': col_bank = i
            elif 'посаж' in s: col_seated = i
            elif 'тех' in s and 'актив' not in s: col_tech = i
        if len(hdr_row) > 14 and 'актив' in str(hdr_row[14] or '').lower():
            col_activation = 14
        else:
            for i, h in enumerate(hdr_row):
                s = str(h or '').lower()
                if col_activation < 0 and 'дата' in s and 'актив' in s and 'цена' not in s:
                    col_activation = i
        if col_activation < 0: col_activation = 14

    def safe(row, i):
        return row[i] if i < len(row) else None

    amt_raw = safe(row, col_amt)
    if amt_raw is None or str(amt_raw).strip() == '':
        amt_raw = safe(row, 9)
    amount = parse_num(amt_raw)
    if not amount or amount <= 0: return None
    paid_at = parse_date(safe(row, col_date), cfg['date_corr'])
    if not paid_at:
        paid_at = f"2025-{month_idx+1:02d}-01"
    # force 2025
    if not paid_at.startswith('2025-'):
        paid_at = '2025-' + paid_at[5:]
    company = str(safe(row, col_client) or '').strip()
    if not company: return None
    cat_raw = str(safe(row, col_cat) or '').strip()
    act_date = parse_date(safe(row, col_activation), cfg['date_corr'])
    if not act_date and col_activation == 14:
        act_date = parse_date(safe(row, 15), cfg['date_corr'])
    period_n = parse_num(safe(row, 8))
    period = max(0, min(60, int(round(period_n)))) if period_n is not None else None
    seated_v = str(safe(row, col_seated) or '').strip()
    seated = seated_v in ('Да','да','+')
    return {
        'paid_at': paid_at,
        'company_name': company,
        'category_raw': cat_raw,
        'category': map_category(cat_raw),
        'amount': amount,
        'amount_planned': parse_num(safe(row, 9)),
        'manager_name': (str(safe(row, col_mgr) or '').strip() or None),
        'bank': (str(safe(row, col_bank) or '').strip() or None),
        'seated': seated,
        'tech_support': (str(safe(row, col_tech) or '').strip() or None),
        'qty': parse_num(safe(row, 4)),
        'price': parse_num(safe(row, 7)),
        'period_months': period,
        'activation_date': act_date,
        'period_start_raw': (str(safe(row, 19)).strip() if safe(row, 19) else None),
        'sheet_tab': month_name,
        'sheet_row': sheet_row_abs,
    }


def sql_v(v):
    if v is None: return 'NULL'
    if isinstance(v, bool): return 'true' if v else 'false'
    if isinstance(v, (int, float)):
        # avoid trailing .0 for ints we stored as float
        if float(v).is_integer() and abs(v) < 1e15:
            return str(int(v))
        return repr(v)
    s = str(v).replace("'", "''")
    return "'" + s + "'"


def main():
    all_rows = []
    stats = {'by_country': {}, 'by_month': {}}
    for country, cfg in COUNTRIES.items():
        stats['by_country'][country] = {'rows': 0, 'sum': 0}
        for mi, month in enumerate(MONTHS):
            f = os.path.join(SRC_DIR, f'{country.lower()}_{month}.json')
            if not os.path.exists(f):
                print(f"MISS {f}", file=sys.stderr); continue
            try:
                with open(f, 'r', encoding='utf-8') as fh:
                    data = json.load(fh)
            except Exception as e:
                print(f"JSON ERR {f}: {e}", file=sys.stderr); continue
            rows = data.get('rows') or []
            if len(rows) < 2: continue
            # find header row
            hdr_idx = -1
            for ri, r in enumerate(rows):
                if r and len(r) > 1 and str(r[1] or '').strip() == 'Компания':
                    hdr_idx = ri; break
            hdr = rows[hdr_idx] if hdr_idx >= 0 else None
            start = hdr_idx + 1 if hdr_idx >= 0 else 4
            m_rows, m_sum = 0, 0
            for idx, row in enumerate(rows[start:]):
                sheet_row_abs = start + idx + 1
                p = parse_row(row, hdr, cfg, month, mi, sheet_row_abs)
                if not p: continue
                p['country'] = country
                p['currency'] = cfg['currency']
                p['source'] = 'sheets_import'
                p['sheet_id'] = cfg['sheet_id']
                all_rows.append(p)
                m_rows += 1
                m_sum += p['amount']
            stats['by_country'][country]['rows'] += m_rows
            stats['by_country'][country]['sum'] += m_sum
            stats['by_month'][f"{country}_{month}"] = {'rows': m_rows, 'sum': m_sum}

    # emit SQL
    cols = ['country','paid_at','company_name','category','category_raw','amount','amount_planned',
            'currency','qty','price','period_months','bank','manager_name','tech_support','seated',
            'activation_date','period_start_raw','source','sheet_id','sheet_tab','sheet_row']
    BATCH = 100
    out = []
    out.append(f"-- Auto-generated 2025 payments import (KZ+KG)")
    out.append(f"-- Total rows: {len(all_rows)}")
    out.append(f"-- Per-country: {json.dumps(stats['by_country'], ensure_ascii=False)}")
    out.append('')
    out.append('BEGIN;')
    out.append('')
    for i in range(0, len(all_rows), BATCH):
        slc = all_rows[i:i+BATCH]
        out.append(f"INSERT INTO payments ({', '.join(cols)}) VALUES")
        vals = []
        for p in slc:
            vs = [sql_v(p[c]) for c in cols]
            vals.append('  (' + ', '.join(vs) + ')')
        out.append(',\n'.join(vals))
        out.append('  ON CONFLICT DO NOTHING;')
        out.append('')
    out.append('COMMIT;')
    out.append('')
    with open(OUT_SQL, 'w', encoding='utf-8') as fh:
        fh.write('\n'.join(out))
    with open(OUT_STATS, 'w', encoding='utf-8') as fh:
        json.dump({'total_rows': len(all_rows), **stats}, fh, ensure_ascii=False, indent=2)
    print(f"OK rows={len(all_rows)} file={OUT_SQL}")
    print(f"stats={json.dumps(stats['by_country'], ensure_ascii=False)}")

if __name__ == '__main__':
    main()
